HIP/hip-801.md

```diff
@@ -0,0 +1,895 @@
+                "memory": [
+                    "0x0000000000000000000000000000000000000000000000000000000000000000",
+                    "0x0000000000000000000000000000000000000000000000000000000000000000",
+                    "0x0000000000000000000000000000000000000000000000000000000000000080"
+                ],
+                "storage": {
+                    "0x0000000000000000000000000000000000000000000000000000000000000000": "0x0000000000000000000000000000000000000000000000000000000000000014"
+                },
+                "reason": null
+            }
+        ]
+}
+```
+
+For providing output formatted by `opcodeLogger`, the Mirror node will re-execute the transaction using the state from the `contract_state_changes` sidecars produced by the consensus nodes. In this way, we can have a track on all of the storage/memory information and the entire trace of opcodes that were executed during the replay.
+
+⚠️ The output generated from the opcodeLogger is very verbose. The /opcodes endpoint will be disabled by default and can be enabled by users of the Hedera Local node or users running their own Mirror Nodes.
+
+### Performance considerations
+
+Since the output generated by this tracer might be quite large, we could add a `flag` for enabling different type of tracers, based on the usage and setup of the Mirror Archive Node. This would give us control and flexibility. A user would be able to enable `opcodeLogger` in a local environment and get the full response data. The default tracer enabled by default on production, could be the `callTracer`.
+
+Since this API would use a lot of resources like memory, CPU usage or requests to the DB, we can have a very strict rate limit that can be configurable based on the context of usage. We can use 1 call per second as initial rate limit. If a given provider want to run the Archive Node with more resources, the rate limit could be increased.
+
+## Backward Compatibility
+
+Mirror Nodes that will support the new REST APIs should have enabled importing of CONTRACT_BYTECODE , CONTRACT_STATE_CHANGE and CONTRACT_ACTION sidecar types. Otherwise, executing the endpoints would result in missing runtime bytecode for execution or using stale contract storage data. In addition, the mirror node should have full historical support (keeping state from the first block), so that it could replay transactions from the entire lifetime of the system.
+
+## Security Implications
+
+There would be some security implementations for Mirror Nodes. Since the debug_trace RPC calls would be free of charge and the Mirror Nodes don’t have throttle mechanism, some attack vectors would be possible. The Nodes could be overloaded with huge amount of calls or calls invoking smart contract methods with huge gas usage, both of which might take more system usage and slow down the network for other users. A rate limit mechanism will be implemented, so that the load put on Mirror Nodes to be balanced.
```

Comment:
This is what we meant, that we don't have a gas based throttle. I will elaborate more on this in the description.

Since the initial purpose of the most heavy API, which is the `/opcodes` one would be mainly for local usage, probably we shouldn't put any limits on the response size for the moment.

I believe the size of the response might increase quicker than the gas increase, since for the same gas cost (executing the same opcode like MLOAD or SLOAD) we can load data with different size, depending on the contracts. So if we decide to have a limit mechanism, apart from the rate limit, it should be based on the response size itself.